modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/idempotency/NuxeoIdempotentFilter.java

```diff
@@ -0,0 +1,196 @@
+        if (kvs == null) {
+            log.debug("KeyValueService not present");
+            return false;
+        }
+        KeyValueStore store = kvs.getKeyValueStore(getStoreName());
+        String storeStatus = store.getString(key + INFO_SUFFIX);
+        byte[] storeContent = store.get(key);
+        if (storeStatus == null) {
+            log.debug("Handle new request for key: {}", key);
+            long ttl = getTTL().toSeconds();
+            store.put(key + INFO_SUFFIX, INPROGRESS_MARKER, ttl);
+            try {
+                try (CopyingResponseWrapper wrapper = new CopyingResponseWrapper(DEFERRED_OUTPUT_STREAM_THRESHOLD,
+                        response)) {
+                    chain.doFilter(request, wrapper);
+                    wrapper.flushBuffer();
+                    byte[] content = wrapper.getCaptureAsBytes();
+                    wrapper.setHeader(HEADER_KEY, key);
+                    if (content.length > MAX_CONTENT_SIZE) {
+                        log.debug(
+                                "Not storing response for key: {} (status: {}, size in bytes: {}), max content size exceeded: {}",
+                                key, wrapper.getStatus(), content.length, MAX_CONTENT_SIZE);
+                        return true;
+                    }
+                    store.put(key, content, ttl);
+                    store.put(key + INFO_SUFFIX, NuxeoIdempotentResponse.write(wrapper), ttl);
+                    log.debug("Stored response for key: {} (status: {}, size in bytes: {})", key, wrapper.getStatus(),
+                            content.length);
+                    return true;
+                }
+            } catch (IOException | ServletException e) {
+                response.setHeader(HEADER_KEY, key);
+                response.setStatus(SC_INTERNAL_SERVER_ERROR);
+                throw e;
+            } finally {
+                if (response.getStatus() >= SC_BAD_REQUEST) {
+                    // error request: cleanup store
+                    store.put(key, (String) null);
+                    store.put(key + INFO_SUFFIX, (String) null);
+                    log.debug("Cleanup store: error for key: {}", key);
+                }
+            }
+        } else if (INPROGRESS_MARKER.equals(storeStatus)) {
+            // request already in progress -> conflict
+            // Don't call response.sendError, because it commits the response
+            // which prevents NuxeoExceptionFilter from returning a custom error page.
+            response.setStatus(SC_CONFLICT);
+            response.setHeader(HEADER_KEY, key);
+            log.debug("Conflict response for key: {}", key);
+            return true;
+        } else {
+            try {
+                // request already done: return stored result
+                NuxeoIdempotentResponse.restore(response, storeStatus.getBytes());
+                response.getOutputStream().write(storeContent);
+                response.setHeader(HEADER_KEY, key);
```

Comment:
You have to set the header before sending the body, because in most cases (unless there's buffering, or in mocked classes) sending the body "commits" the response (see `ServletResponse.isCommitted`) and you can't set any header or status after that.